pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java

```diff
@@ -75,25 +75,35 @@ public static List<ImmutableTriple<MessageId, String, Integer>> extractIdsAndKey
                                                   msg.getMessageIdData().getEntryId(),
                                                   msg.getMessageIdData().getPartition(),
                                                   i);
-            if (!smm.isCompactedOut() && smm.hasPartitionKey()) {
-                idsAndKeysAndSize.add(ImmutableTriple.of(id,
-                        smm.getPartitionKey(),
-                        smm.hasPayloadSize() ? smm.getPayloadSize() : 0));
+            if (!smm.isCompactedOut()) {
+                if (smm.hasPartitionKey()) {
+                    idsAndKeysAndSize.add(ImmutableTriple.of(id,
+                            smm.getPartitionKey(),
+                            smm.hasPayloadSize() ? smm.getPayloadSize() : 0));
+                } else {
+                    idsAndKeysAndSize.add(null);
```

Comment:
please avoid add null to the list. :)

Reply:
+1
https://github.com/apache/pulsar/blob/2322004069f23ecca1a12a4ced898fed854275fb/pulsar-broker/src/main/java/org/apache/pulsar/compaction/TwoPhaseCompactor.java#L135-L147

Reply:
Fixed.